dev/system-test/tracing.ts

```diff
@@ -402,218 +611,235 @@ describe('Tracing Tests', () => {
+  describe(E2E_TEST_SUITE_TITLE, () => {
+    describe(NON_GLOBAL_OTEL_TEST_SUITE_TITLE, () => {
+      describe(GRPC_TEST_SUITE_TITLE, () => {
+        beforeEach(function () {
+          beforeEachTest(this.currentTest!);
+        });
+        runTestCases();
         afterEach(async () => afterEachTest());
       });
-      describe('REST', () => {
-        const config: TestConfig = {
-          e2e: true,
-          useGlobalOpenTelemetry: false,
-          preferRest: true,
-        };
-        beforeEach(async () => beforeEachTest(config));
-        runTestCases(config);
+      describe(REST_TEST_SUITE_TITLE, () => {
+        beforeEach(function () {
+          beforeEachTest(this.currentTest!);
+        });
+        runTestCases();
         afterEach(async () => afterEachTest());
       });
     });
-    describe('with Global-OTEL', () => {
-      describe('GRPC', () => {
-        const config: TestConfig = {
-          e2e: true,
-          useGlobalOpenTelemetry: true,
-          preferRest: false,
-        };
-        beforeEach(async () => beforeEachTest(config));
-        runTestCases(config);
+    describe(GLOBAL_OTEL_TEST_SUITE_TITLE, () => {
+      describe(GRPC_TEST_SUITE_TITLE, () => {
+        beforeEach(function () {
+          beforeEachTest(this.currentTest!);
+        });
+        runTestCases();
         afterEach(async () => afterEachTest());
       });
-      describe('REST', () => {
-        const config: TestConfig = {
-          e2e: true,
-          useGlobalOpenTelemetry: true,
-          preferRest: true,
-        };
-        beforeEach(async () => beforeEachTest(config));
-        runTestCases(config);
+      describe(REST_TEST_SUITE_TITLE, () => {
+        beforeEach(function () {
+          beforeEachTest(this.currentTest!);
+        });
+        runTestCases();
         afterEach(async () => afterEachTest());
       });
     });
   });
 
-  function runTestCases(config: TestConfig) {
+  function runTestCases() {
     it('document reference get()', async () => {
-      await firestore.collection('foo').doc('bar').get();
-
-      await waitForCompletedSpans(config, 2);
-      expectSpanHierarchy(SPAN_NAME_DOC_REF_GET, SPAN_NAME_BATCH_GET_DOCUMENTS);
+      await runFirestoreOperationInRootSpan(() =>
+        firestore.collection('foo').doc('bar').get()
+      );
+      await waitForCompletedSpans(3);
+      expectSpanHierarchy(
+        SPAN_NAME_TEST_ROOT,
+        SPAN_NAME_DOC_REF_GET,
+        SPAN_NAME_BATCH_GET_DOCUMENTS
+      );
     });
 
     it('document reference create()', async () => {
-      await firestore.collection('foo').doc().create({});
-
-      await waitForCompletedSpans(config, 2);
-      expectSpanHierarchy(SPAN_NAME_DOC_REF_CREATE, SPAN_NAME_BATCH_COMMIT);
+      await runFirestoreOperationInRootSpan(() =>
+        firestore.collection('foo').doc().create({})
+      );
+      await waitForCompletedSpans(3);
+      expectSpanHierarchy(
+        SPAN_NAME_TEST_ROOT,
```

Comment:
is the added span specific to testing or the tracing added root layer span generally? What is the reason behind wrapping the operations with `runFirestoreOperationInRootSpan`and adding this extra root span?

Reply:
Great question. Many tracing tests are run in a short period of time, and each test sends one trace and potentially several spans to the same tracing backend (cloud trace).

We want each test to be able to query Cloud Trace to find out if the expected spans were reported for that test. In order to distinguish between the different tests, we need to do two things
1- create a custom context (which contains a unique Trace ID) for each test (done in `getNewSpanContext`)
2- create a root span _with that context_ so that all Firestore spans for that test fall under this root span.
3- query the cloud trace backend for that unique Trace ID.

doing (1) without (2) doesn't work.